jina/peapods/pods/compoundpod.py

```diff
@@ -0,0 +1,270 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+import argparse
+from argparse import Namespace
+from typing import Optional, Dict, List, Union, Set
+
+from .. import Pea
+from .. import Pod
+from .. import BasePod
+from ...enums import PollingType, PeaRoleType
+
+
+class CompoundPod(BasePod):
+    """A CompoundPod is a immutable set of pods, which run in parallel. They share the same input and output socket.
+    Internally, the peas of the pods can run with the process/thread backend. They can be also run in their own containers.
+    :param args: pod arguments parsed from the CLI
+    :param needs: pod names of preceding pods, the output of these pods are going into the input of this pod
+    """
+
+    def __init__(self, args: Union['argparse.Namespace', Dict], needs: Set[str] = None):
+        super().__init__(args, needs)
+        self.replica_list = []  # type: List['Pod']
+        if isinstance(args, Dict):
+            # This is used when a Pod is created in a remote context, where peas & their connections are already given.
+            self.replicas_args = args
+        else:
+            self.replicas_args = self._parse_args(args)
+
+    @property
+    def first_pod_args(self) -> Namespace:
+        """
+        Return the first non-head/tail pod args
+
+        :return: arguments of the first pod
+        """
+        # note this will be never out of boundary
+        return self.replicas_args['replicas'][0]
+
+    @property
+    def port_expose(self) -> int:
+        """Get the grpc port number
+
+
+        .. # noqa: DAR201
+        """
+        return self.first_pod_args.port_expose
+
+    @property
+    def host(self) -> str:
+        """Get the host name of this Pod
+
+
+        .. # noqa: DAR201
+        """
+        return self.first_pod_args.host
+
+    def _parse_args(
+        self, args: Namespace
+    ) -> Dict[str, Optional[Union[List[Namespace], Namespace]]]:
+        return self._parse_base_pod_args(
+            args,
+            attribute='replicas',
+            id_attribute_name='replica_id',
+            role_type=PeaRoleType.REPLICA,
+            repetition_attribute='replicas',
+            polling_type=PollingType.ANY,
+        )
+
+    @property
+    def head_args(self):
+        """
+        Get the arguments for the `head` of this BasePod.
+
+        :return: arguments of the head pea
+        """
+        if self.is_head_router and self.replicas_args['head']:
```

Comment:
I think that for `CompoundPod` we need to assume that THERE ABSOLUTELY is a `head Pea` in front of them, so for now not need to do all these IF ELSE...

Reply:
Same applies to `tailPea`

Reply:
nice catch! I changed it